_posts/2019-06-06-NeuralODEs.markdown

```diff
@@ -0,0 +1,388 @@
+
+Finally, thank you to all my fellow students who attended the recitations and provided valuable feedback.
+
+<div class="deps-graph">
+<iframe class="deps" src="/assets/nodes-deps.svg" width="200"></iframe>
+<div>Concepts used in Neural ODEs. Click to navigate.</div>
+</div>
+
+# Why
+
+Neural ODEs are neural network models which generalize standard layer to layer propagation to continuous depth models. Starting from the observation that the forward propagation in neural networks is equivalent to one step of discretation of an ODE, we can construct and efficiently train models by via ODEs. On top of providing a novel family of architectures, notably for invertible density models and continuous time series, neural ODEs also provide a memory efficiency gain in supervised learning tasks.
+
+In this curriculum, we will go through all the background topics necessary to understand these models. At the end, you should be able to implement neural ODEs and apply them to different tasks.
+
+<br />
+
+# Common resources:
+
+1. (Süli & Mayers): [An Introduction to Numerical Analysis](https://www.cambridge.org/core/books/an-introduction-to-numerical-analysis/FD8BCAD7FE68002E2179DFF68B8B7237#)
+2. (Quarteroni et al.): [Numerical Mathematics](https://www.springer.com/us/book/9783540346586?token=holiday18&utm_campaign=3_fjp8312_us_dsa_springer_holiday18&gclid=Cj0KCQiAvebhBRD5ARIsAIQUmnlViB7VsUn-2tABSAhIvYaJgSEqmJXD7F4A7EgyDQtY9v_GeUsNif8aArGAEALw_wcB)
+
+# 1 Numerical solution of ODEs - Part 1
+  **Motivation**: ODEs are used to mathematically model a number of natural processes and phenomena. The study of their numerical 
+    simulations is one of the main topics in numerical analysis and of fundamental importance in applied sciences. To understand Neural ODEs, we need to first understand how ODEs are solved with numerical techniques.
+
+  **Topics**:
+
+  1. Initial values problems
+  2. One-step methods
+  3. Consistency and convergence
+
+  **Required Reading**:
+
+  1. Sections 12.1-4 from (Süli & Mayers)
+  2. Sections 11.1-3 from (Quarteroni et al.)
+  
+  **Notes**: Here is a [link](/assets/nodes_notes/week1.pdf) to our notes for the lesson.
+  
+  **Optional Reading**:
+
+  1. Runge-Kutta methods: Section 12.5 from (Süli & Mayers)
+  2. [Prof. Trefethen's class ODEs and Nonlinear Dynamics 4.2](http://podcasts.ox.ac.uk/odes-and-nonlinear-dynamics-42)
+
+  **Questions**:
+
+  1. Exercise 1 in Section 11.12 of (Quarteroni et al.)
+     <details><summary>Solution</summary>
+     <p>
+     Start from the hint in the book and notice that you can bound \(E_2\) as
+     
+     $$|E_2|\leq \frac{hL}{2}|y_{n+1}-y_{n} - hf(t_n,y_n)| = \frac{hL}{2}O(h^2) = O(h^3)$$
+     
+     where \(L\) is the Lipschitz constant of \(f\). On the other hand, \(E_1 = O(h**3)\) (see [link](https://en.wikipedia.org/wiki/Trapezoidal_rule#Error_analysis)).
+     </p>
+     </details>
+
+  2. Exercises 12.3,12.4, 12.7 in Section 12 of (Süli & Mayers)
```

Comment:
Exercise 2 solution:
	- What’s up with the spacing in 12.3? Can we fix that?
	- In 12.4, the word “Instead” isn’t right. I’m not really sure what it should be though because the solution also seems to end abruptly.

Reply:
fixed - the solution is pretty much complete, one only needs to substitute C_n in the equation and solve some boring algebra.

Reply:
- 12.3 still has really weird spacing. Do you see what I mean on the 4th, 5th, and 6th lines?
- 12.4: "we see that it holds y(nh) = 0 for all n >= 0" --> "we see that y(nh) = 0 for all n >= 0".
- 12.4: "method reads" --> "method says that"

Reply:
fixed